src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java

```diff
@@ -0,0 +1,34 @@
+package org.mozartoz.truffle.nodes.call;
+
+import org.mozartoz.truffle.nodes.OzNode;
+import org.mozartoz.truffle.runtime.TailCallException;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.profiles.BranchProfile;
+
+public class TailCallCatcherNode extends OzNode {
+
+	@Child CallNode callNode;
+
+	public TailCallCatcherNode(CallNode callNode) {
+		this.callNode = callNode;
+	}
+
+	BranchProfile tailCallProfile = BranchProfile.create();
+
+	public Object execute(VirtualFrame frame) {
+		try {
+			return callNode.execute(frame);
+		} catch (TailCallException e) {
+			tailCallProfile.enter();
+			while (true) {
+				try {
+					return callNode.executeCall(frame, e.receiver, e.arguments);
+				} catch (TailCallException exception) {
+					e = exception;
```

Comment:
Reassigning a catch variable is not too nice, could you make a tailCall local var?